File: examples/php/fob-simple/advanced_example.php
#!/usr/bin/env php
<?php
/**
 * Advanced Fob Bundler Example
 *
 * Demonstrates advanced features like:
 * - Library builds with external dependencies
 * - App builds with code splitting
 * - Component library builds
 * - Custom configuration options
 */

// Load the native extension
if (!extension_loaded('fob')) {
    echo "❌ Error: fob extension not found!\n";
    echo "Build and install the extension first:\n";
    echo "  cargo build --release --package fob-php\n";
    echo "  cargo php install --release\n";
    exit(1);
}

function example_library() {
    echo "📚 Example 1: Library Build (externalize dependencies)\n";
    echo str_repeat('-', 60) . "\n";
    
    try {
        // Build a library - dependencies are externalized
        $result = fob_library(
            'src/index.js',
            [
                'out_dir' => 'dist/library',
                'format' => 'esm',
                'external' => ['react', 'react-dom'], // Externalize these packages
                'sourcemap' => 'external',
            ]
        );
        
        echo "✅ Library build complete!\n";
        echo "   Chunks: " . count($result['chunks']) . "\n";
        echo "   Modules: {$result['stats']['total_modules']}\n\n";
        
    } catch (Exception $e) {
        echo "❌ Failed: {$e->getMessage()}\n\n";
    }
}

function example_app() {
    echo "🚀 Example 2: App Build (with code splitting)\n";
    echo str_repeat('-', 60) . "\n";
    
    try {
        // Build an app with code splitting
        $result = fob_app(
            ['src/index.js', 'src/utils.js'], // Multiple entries
            [
                'out_dir' => 'dist/app',
                'format' => 'esm',
                'code_splitting' => [
                    'min_size' => 20000,
                    'min_imports' => 2,
                ],
                'minify' => true,
            ]
        );
        
        echo "✅ App build complete!\n";
        echo "   Chunks: " . count($result['chunks']) . "\n";
        echo "   Total size: {$result['stats']['total_size']} bytes\n\n";
        
    } catch (Exception $e) {
        echo "❌ Failed: {$e->getMessage()}\n\n";
    }
}

function example_components() {
    echo "🧩 Example 3: Component Library Build\n";
    echo str_repeat('-', 60) . "\n";
    
    try {
        // Build a component library - each entry is isolated
        $result = fob_components(
            ['src/index.js', 'src/utils.js'],
            [
                'out_dir' => 'dist/components',
                'format' => 'esm',
                'external_from_manifest' => true, // Externalize from package.json
            ]
        );
        
        echo "✅ Component library build complete!\n";
        echo "   Chunks: " . count($result['chunks']) . "\n";
        echo "   Module count: {$result['module_count']}\n\n";
        
    } catch (Exception $e) {
        echo "❌ Failed: {$e->getMessage()}\n\n";
    }
}

function example_custom_config() {
    echo "⚙️  Example 4: Custom Configuration\n";
    echo str_repeat('-', 60) . "\n";
    
    try {
        // Use the Fob class for full control
        $bundler = new Fob([
            'entries' => ['src/index.js'],
            'output_dir' => 'dist/custom',
            'format' => 'esm',
            'sourcemap' => 'inline',
            'platform' => 'browser',
            'minify' => false,
            'cwd' => __DIR__,
        ]);
        
        $result = $bundler->bundle();
        
        echo "✅ Custom build complete!\n";
        echo "   Format: esm\n";
        echo "   Sourcemap: inline\n";
        echo "   Platform: browser\n";
        echo "   Duration: {$result['stats']['duration_ms']}ms\n\n";
        
    } catch (Exception $e) {
        echo "❌ Failed: {$e->getMessage()}\n\n";
    }
}

function main() {
    echo "🎯 Advanced Fob Examples\n\n";
    
    // Initialize logging
    fob_init_logging('info');
    
    // Run examples
    example_library();
    example_app();
    example_components();
    example_custom_config();
    
    echo "✨ All examples complete!\n";
}

main();

File: examples/python/fob-simple/advanced_example.py
#!/usr/bin/env python3
"""
Advanced Fob Bundler Example

Demonstrates more advanced features:
- Multiple entry points
- Code splitting
- Library mode
- Component library mode
- Custom configuration
"""

import asyncio
import sys
from pathlib import Path

try:
    import fob
except ImportError:
    print("❌ Error: fob module not found!")
    print("\nTo install:")
    print("  cd ../../..")
    print("  maturin develop --manifest-path crates/fob-python/Cargo.toml")
    sys.exit(1)


async def example_bundle_entry():
    """Example: Bundle a single entry point"""
    print("📦 Example 1: Bundle Entry")
    print("-" * 40)
    
    result = await fob.Fob.bundle_entry(
        "src/index.js",
        {
            "out_dir": "dist/bundle-entry",
            "format": "esm",
            "minify": True,
            "sourcemap": "external"
        }
    )
    
    print(f"✅ Bundled {result['stats']['total_modules']} modules")
    print(f"   Output: {result['chunks'][0]['file_name']}\n")


async def example_library():
    """Example: Build a library (externalize dependencies)"""
    print("📚 Example 2: Library Mode")
    print("-" * 40)
    
    result = await fob.Fob.library(
        "src/index.js",
        {
            "out_dir": "dist/library",
            "external": ["react", "react-dom"],  # Can also be a single string
            "format": "esm"
        }
    )
    
    print(f"✅ Library built")
    print(f"   Externalized dependencies from package.json\n")


async def example_app():
    """Example: Build an app with code splitting"""
    print("🌐 Example 3: App Mode (Code Splitting)")
    print("-" * 40)
    
    # Using pathlib.Path for entries
    entries = [Path("src/index.js")]
    
    result = await fob.Fob.app(
        entries,
        {
            "out_dir": "dist/app",
            "code_splitting": {
                "min_size": 1000,  # Minimum chunk size in bytes
                "min_imports": 2   # Minimum shared imports
            },
            "format": "esm"
        }
    )
    
    print(f"✅ App built with code splitting")
    print(f"   Generated {result['stats']['total_chunks']} chunks")
    print(f"   Shared dependencies extracted\n")


async def example_components():
    """Example: Build a component library"""
    print("🧩 Example 4: Component Library")
    print("-" * 40)
    
    result = await fob.Fob.components(
        ["src/index.js"],  # In real usage: ["src/Button.tsx", "src/Card.tsx"]
        {
            "out_dir": "dist/components",
            "format": "esm",
            "external_from_manifest": True
        }
    )
    
    print(f"✅ Component library built")
    print(f"   Each entry produces independent bundle\n")


async def example_custom_config():
    """Example: Full custom configuration"""
    print("⚙️  Example 5: Custom Configuration")
    print("-" * 40)
    
    bundler = fob.Fob({
        "entries": ["src/index.js"],
        "output_dir": "dist/custom",
        "format": "esm",
        "platform": "browser",  # or "node"
        "sourcemap": "inline",
        "minify": False,
        "external": ["lodash"],  # Externalize specific packages
        "cwd": Path.cwd()  # Can use pathlib.Path
    })
    
    result = await bundler.bundle()
    
    print(f"✅ Custom build complete")
    print(f"   Platform: browser")
    print(f"   Sourcemap: inline")
    print(f"   External packages: lodash\n")


async def example_error_handling():
    """Example: Error handling"""
    print("⚠️  Example 6: Error Handling")
    print("-" * 40)
    
    try:
        # This will fail - file doesn't exist
        result = await fob.bundle_single(
            "nonexistent.js",
            "dist",
            "esm"
        )
    except fob.FobError as e:
        print(f"✅ Caught FobError: {e}\n")
    except Exception as e:
        print(f"❌ Unexpected error: {e}\n")


async def main():
    """Run all examples"""
    print("🚀 Advanced Fob Python Examples\n")
    print("=" * 50 + "\n")
    
    # Initialize logging
    fob.init_logging("info")
    
    try:
        await example_bundle_entry()
        await example_library()
        await example_app()
        await example_components()
        await example_custom_config()
        await example_error_handling()
        
        print("=" * 50)
        print("✅ All examples completed!")
        print("\n💡 Check the dist/ directory for outputs")
        
    except Exception as e:
        print(f"\n❌ Example failed: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
